File: src/chain-observers/collectors/collector-base.ts
import type { WorkerBee } from "../../bot";

import type * as TClassifiers from "../classifiers";
import { TRegisterEvaluationContext } from "../classifiers/collector-classifier-base";
import { DataEvaluationContext } from "../factories/data-evaluation-context";

export type TAvailableClassifiers = {
  [key in keyof typeof TClassifiers]: InstanceType<(typeof TClassifiers)[key]>["type"];
};

export class CollectorBase {
  public constructor(
    protected readonly worker: WorkerBee
  ) {}

  private registersCount = 0;

  /**
   * We need to return the data in the format of { [classifierName]: { [key]: value } } so
   * overriding can be properly deduced later by the data evaluation context
   *
   * This method allows to get collector state and cache the request, which can be used in the evaluation context {@link DataEvaluationContext.get} method
   * Think of this method as HTTP GET request, while {@link CollectorBase.query} is like HTTP POST request
   *
   * When user calls this method via Data evaluation context, and it is not overridden by the subclass, it will fail with an error
   */
  public get?(data: DataEvaluationContext): Promise<Partial<TAvailableClassifiers>>;

  /**
   * We need to return the data in the format of { [classifierName]: { [key]: value } } so
   * overriding can be properly deduced later by the data evaluation context
   *
   * This method allows to query the collector for data (without cache), which can be used in the evaluation context {@link DataEvaluationContext.query} method
   * Think of this method as HTTP POST request, while {@link CollectorBase.get} is like HTTP GET request
   *
   * When user calls this method via Data evaluation context, and it is not overridden by the subclass, it will fail with an error
   */
  public query?(data: DataEvaluationContext, options: object): Promise<Partial<TAvailableClassifiers>>;

  /**
   * If the collector uses any context, it should report it as dependency by overriding this function
   */
  public usedContexts(): Array<TRegisterEvaluationContext> {
    return [];
  }

  protected pushOptions?(data: Record<string, any>): void;
  protected popOptions?(data: Record<string, any>): void;

  public get hasRegistered() {
    return this.registersCount > 0;
  }

  public register(data?: Record<string, any>) {
    ++this.registersCount;

    if (data !== undefined)
      this.pushOptions?.(data);
  }
  public unregister(data?: Record<string, any>) {
    --this.registersCount;

    if (data !== undefined)
      this.popOptions?.(data);
  }
}

File: python/workerbee/chain_observers/collectors/collector_base.py
"""CollectorBase — base class for all collectors.

Mirrors src/chain-observers/collectors/collector-base.ts.
Provides reference counting via register/unregister and optional get/query methods.
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any

if TYPE_CHECKING:
    from ..classifiers.collector_classifier_base import TRegisterEvaluationContext
    from ..factories.data_evaluation_context import TCollectorEvaluationContext
    from ..interfaces import IWorkerBee


class CollectorBase:
    """Base class for all collectors.

    Subclasses override get() and/or query() to provide data.
    The factory manages registration via push_classifier/pop_classifier which
    calls register/unregister here for reference counting.
    """

    def __init__(self, worker: IWorkerBee) -> None:
        self.worker = worker
        self._registers_count: int = 0

    @property
    def has_registered(self) -> bool:
        return self._registers_count > 0

    def used_contexts(self) -> list[TRegisterEvaluationContext]:
        return []

    async def get(self, data: TCollectorEvaluationContext) -> dict[str, Any]:
        raise NotImplementedError(
            f"Collector {type(self).__name__} does not implement the requested 'get' method",
        )

    async def query(self, data: TCollectorEvaluationContext, options: Any) -> Any:
        raise NotImplementedError(
            f"Collector {type(self).__name__} does not implement the requested 'query' method",
        )

    def push_options(self, data: Any) -> None:
        pass

    def pop_options(self, data: Any) -> None:
        pass

    def register(self, data: Any = None) -> None:
        self._registers_count += 1
        if data is not None:
            self.push_options(data)

    def unregister(self, data: Any = None) -> None:
        self._registers_count -= 1
        if data is not None:
            self.pop_options(data)
